express_relay/sdk/python/express_relay_utils/express_relay_client.py

```diff
@@ -0,0 +1,404 @@
+from eth_abi import encode
+from eth_account.account import Account
+from web3.auto import w3
+from express_relay_utils.express_relay_types import (
+    Opportunity,
+    BidStatusUpdate,
+    ClientMessage,
+    BidStatus,
+    Bid,
+    OpportunityBid,
+    OpportunityParams,
+)
+
+
+class ExpressRelayClientException(Exception):
+    pass
+
+
+class ExpressRelayClient:
+    def __init__(
+        self,
+        server_url: str,
+        opportunity_callback: (
+            Callable[[Opportunity], Coroutine[Any, Any, Any]] | None
+        ) = None,
+        bid_status_callback: (
+            Callable[[BidStatusUpdate], Coroutine[Any, Any, Any]] | None
+        ) = None,
+        **kwargs,
+    ):
+        """
+        Args:
+            server_url: The URL of the auction server.
+            opportunity_callback: An async function that serves as the callback on a new opportunity. Should take in one external argument of type Opportunity.
+            bid_status_callback: An async function that serves as the callback on a new bid status update. Should take in one external argument of type BidStatusUpdate.
+            kwargs: Keyword arguments to pass to the websocket connection.
+        """
+        parsed_url = urllib.parse.urlparse(server_url)
+        if parsed_url.scheme == "https":
+            ws_scheme = "wss"
+        elif parsed_url.scheme == "http":
+            ws_scheme = "ws"
+        else:
+            raise ValueError("Invalid server URL")
+
+        self.server_url = server_url
+        self.ws_endpoint = parsed_url._replace(scheme=ws_scheme, path="/v1/ws").geturl()
+        self.ws_msg_counter = 0
+        self.ws: WebSocketClientProtocol
+        self.ws_lock = asyncio.Lock()
+        self.ws_loop: Task[Any]
+        self.ws_msg_futures: dict[str, asyncio.Future] = {}
+        self.ws_options = kwargs
+        self.opportunity_callback = opportunity_callback
+        self.bid_status_callback = bid_status_callback
+
+    async def start_ws(self):
+        """
+        Initializes the websocket connection to the server, if not already connected.
+        """
+        async with self.ws_lock:
+            if not hasattr(self, "ws"):
+                self.ws = await websockets.connect(self.ws_endpoint, **self.ws_options)
+
+            if not hasattr(self, "ws_loop"):
+                ws_call = self.ws_handler(
+                    self.opportunity_callback, self.bid_status_callback
+                )
+                self.ws_loop = asyncio.create_task(ws_call)
+
+    async def close_ws(self):
+        """
+        Closes the websocket connection to the server.
+        """
+        async with self.ws_lock:
+            await self.ws.close()
+
+    async def get_ws_loop(self) -> asyncio.Task:
+        """
+        Returns the websocket handler loop.
+        """
+        await self.start_ws()
+
+        return self.ws_loop
+
+    async def send_ws_message(self, method: str, params: dict) -> dict:
```

Comment:
I think it makes more sense for this function to accept a ClientMessage and the callers be responsible to craft the correct payloads

Reply:
yeah i wanted to avoid multiple duplicate calls to `ClientMessage.model_validate` throughout the code. but i dont have a strong take on this, so can refactor it